File: src/main/java/com/joecollins/graphics/screens/generic/ProjectionScreen.java
package com.joecollins.graphics.screens.generic;

import com.joecollins.bindings.Binding;
import com.joecollins.bindings.BindingReceiver;
import com.joecollins.graphics.components.ProjectionFrame;
import java.awt.Color;
import java.awt.GridLayout;
import java.awt.Image;
import javax.swing.JPanel;
import javax.swing.border.EmptyBorder;

public class ProjectionScreen extends JPanel {

  public static ProjectionScreen createScreen(
      Binding<String> text, Binding<Color> color, Binding<Image> image) {
    return new ProjectionScreen(
        text, color, image, Binding.fixedBinding(ProjectionFrame.Alignment.BOTTOM));
  }

  public static ProjectionScreen createScreen(
      Binding<String> text,
      Binding<Color> color,
      Binding<Image> image,
      Binding<ProjectionFrame.Alignment> imageAlignment) {
    return new ProjectionScreen(text, color, image, imageAlignment);
  }

  private ProjectionScreen(
      Binding<String> text,
      Binding<Color> color,
      Binding<Image> image,
      Binding<ProjectionFrame.Alignment> imageAlignment) {
    setLayout(new GridLayout(1, 1));
    setBackground(Color.WHITE);
    setBorder(new EmptyBorder(5, 5, 5, 5));

    BindingReceiver<Color> colorReceiver = new BindingReceiver<>(color);
    ProjectionFrame frame = new ProjectionFrame();
    frame.setImageBinding(image);
    frame.setHeaderBinding(() -> "PROJECTION");
    frame.setBackColorBinding(colorReceiver.getBinding());
    frame.setBorderColorBinding(colorReceiver.getBinding());
    frame.setFooterTextBinding(text);
    frame.setImageAlignmentBinding(imageAlignment);
    add(frame);
  }
}

File: src/main/kotlin/com/joecollins/graphics/screens/generic/ProjectionScreen.kt
package com.joecollins.graphics.screens.generic

import com.joecollins.graphics.AltTextProvider
import com.joecollins.graphics.components.ProjectionFrame
import com.joecollins.pubsub.asOneTimePublisher
import com.joecollins.pubsub.map
import java.awt.Color
import java.awt.GridLayout
import java.awt.Image
import java.util.concurrent.Flow
import javax.swing.JPanel
import javax.swing.border.EmptyBorder

class ProjectionScreen private constructor(
    text: Flow.Publisher<out String?>,
    color: Flow.Publisher<out Color>,
    image: Flow.Publisher<out Image?>,
    imageAlignment: Flow.Publisher<out ProjectionFrame.Alignment>
) : JPanel(), AltTextProvider {

    override val altText: Flow.Publisher<String?> = text.map { t -> t?.let { "PROJECTION: $it" } }

    companion object {
        fun createScreen(
            text: Flow.Publisher<out String?>,
            color: Flow.Publisher<out Color>,
            image: Flow.Publisher<out Image?>
        ): ProjectionScreen {
            return ProjectionScreen(
                text,
                color,
                image,
                ProjectionFrame.Alignment.BOTTOM.asOneTimePublisher()
            )
        }

        fun createScreen(
            text: Flow.Publisher<out String?>,
            color: Flow.Publisher<out Color>,
            image: Flow.Publisher<out Image?>,
            imageAlignment: Flow.Publisher<out ProjectionFrame.Alignment>
        ): ProjectionScreen {
            return ProjectionScreen(text, color, image, imageAlignment)
        }
    }

    init {
        layout = GridLayout(1, 1)
        background = Color.WHITE
        border = EmptyBorder(5, 5, 5, 5)
        val frame = ProjectionFrame(
            headerPublisher = "PROJECTION".asOneTimePublisher(),
            borderColorPublisher = color,
            imagePublisher = image,
            backColorPublisher = color,
            footerTextPublisher = text,
            imageAlignmentPublisher = imageAlignment
        )
        add(frame)
    }
}
